File: src/kdsnr_hwp_toolkit/transform/slot_catalog.py
"""Slot catalog — derives atom→(pp,st,cs) mapping from each subject's templet.

Single source of truth: templet/{subject}.hwpx. We walk paragraphs, classify
each, and pick the most-common slot per atom (single-unification rule
confirmed with user 2026-05-11). Box atoms also store the box paragraph
itself for clone-and-swap transforms.
"""
from __future__ import annotations

from collections import Counter, defaultdict
from dataclasses import dataclass
from functools import lru_cache

from kdsnr_hwp_toolkit.codec import read
from kdsnr_hwp_toolkit.codec.schema import (
    CharItem, HwpxDocument, Paragraph, TableItem,
)
from kdsnr_hwp_toolkit.resources import template_bytes
from kdsnr_hwp_toolkit.classify.atoms import Atom, BOX_ATOMS
from kdsnr_hwp_toolkit.classify.classifier import classify


SubjectName = str  # "math" | "science" | "social" | "korean"


@dataclass(frozen=True)
class Slot:
    """A templet slot — paragraph properties + optional reference paragraph.

    body_cs_id: for balmun (and any other atom whose templet paragraph carries
    two distinct char_shapes — first char = number/marker style, rest = body
    style). When set, transform applies cs_id only to the leading "N." run
    and body_cs_id to everything after. None = single cs across the paragraph.
    """

    pp_id: int
    st_id: int
    cs_id: int
    template_paragraph: Paragraph | None = None  # for box atoms (clone source)
    body_cs_id: int | None = None


@dataclass(frozen=True)
class SubjectCatalog:
    subject: SubjectName
    template: HwpxDocument
    atom_to_slot: dict  # Atom → Slot


def _first_table(p: Paragraph) -> TableItem | None:
    for it in p.items:
        if isinstance(it, TableItem):
            return it
    return None


def _is_1x1(tbl: TableItem) -> bool:
    return (
        tbl.table_attrs.get("rowCnt") == "1"
        and tbl.table_attrs.get("colCnt") == "1"
    )


def _has_bogi_marker(tbl: TableItem) -> bool:
    import re as _re
    for cell in tbl.cells:
        text = "".join(
            it.text
            for cp in cell.paragraphs
            for it in cp.items
            if isinstance(it, CharItem)
        )
        if "보기" in _re.sub(r"\s+", "", text):
            return True
    return False


def _prefer_wrapper(atom: Atom, candidates: list[Paragraph]) -> Paragraph:
    """Pick canonical wrapper paragraph for box atoms.

    DATA_BOX / JIMUN_DATA_BOX → 1×1 shell preferred (avoid leaking inline
    table labels like math's '합계' from confusion-matrix wrappers).
    BOGI_BOX → paragraph whose table carries '보기' marker.
    Others → first match.
    """
    if atom in (Atom.DATA_BOX, Atom.JIMUN_DATA_BOX):
        for p in candidates:
            tbl = _first_table(p)
            if tbl is not None and _is_1x1(tbl):
                return p
    if atom == Atom.BOGI_BOX:
        for p in candidates:
            tbl = _first_table(p)
            if tbl is not None and _has_bogi_marker(tbl):
                return p
    return candidates[0]


def _detect_body_cs(p: Paragraph, first_cs: int) -> int | None:
    """Find the first char_shape that differs from first_cs in this paragraph.

    Returns None if every item shares first_cs (= single-cs paragraph).
    Used to detect templet's "marker char + body char" pattern (balmun's
    "N." in number style, body in body style).
    """
    for it in p.items:
        cs = getattr(it, "char_shape_id", None)
        if cs is not None and cs != first_cs:
            return cs
    return None


def _classify_template(template: HwpxDocument, subject: str):
    """Walk template paragraphs, returning [(idx, atom, paragraph), ...]."""
    body = template.sections[0].body
    out = []
    prev = None
    for i, p in enumerate(body):
        atom = classify(p, prev_atom=prev, subject=subject)
        out.append((i, atom, p))
        if atom not in (Atom.EMPTY, Atom.UNKNOWN):
            prev = atom
    return out


def _build_catalog(subject: SubjectName) -> SubjectCatalog:
    raw = template_bytes(f"{subject}.hwpx")
    template = read(raw, doc_id=f"template:{subject}")
    classified = _classify_template(template, subject)

    # Group paragraphs by atom
    by_atom: dict[Atom, list[Paragraph]] = defaultdict(list)
    for _, atom, p in classified:
        if atom in (Atom.EMPTY, Atom.UNKNOWN):
            continue
        by_atom[atom].append(p)

    # For each atom, pick the most common (pp, st, cs) — user confirmed
    # "차이 미미하면 더 많이 쓰인 슬롯으로 단일 통일"
    atom_to_slot: dict[Atom, Slot] = {}
    for atom, paras in by_atom.items():
        triple_counter: Counter = Counter(
            (p.para_shape_id, p.style_id, p.char_shape_id_first) for p in paras
        )
        chosen_triple, _ = triple_counter.most_common(1)[0]

        # Special case for ⑪ (social data_box widths): pick FIRST encountered
        # (user explicit: "첫번째로 통일").
        if atom == Atom.DATA_BOX and subject == "social":
            chosen_triple = (
                paras[0].para_shape_id,
                paras[0].style_id,
                paras[0].char_shape_id_first,
            )

        # BOGI_BOX: prefer slots whose representative is a CLEAN wrapper
        # (table only, no fused balmun-cont text). Otherwise the cloned
        # paragraph carries fused leading text into output → "발문 두 번
        # 반복" artifact.
        if atom == Atom.BOGI_BOX:
            clean_paras = [
                p for p in paras
                if not any(
                    isinstance(it, CharItem) and it.text.strip()
                    for it in p.items
                )
            ]
            if clean_paras:
                clean_triples = Counter(
                    (p.para_shape_id, p.style_id, p.char_shape_id_first)
                    for p in clean_paras
                )
                chosen_triple, _ = clean_triples.most_common(1)[0]

        # Pick a representative paragraph that uses the chosen triple.
        # For box atoms prefer the canonical wrapper shape (1×1 / bogi-marked)
        # — otherwise the first inline-table paragraph would be picked and
        # _box_clone would leak its labels into output.
        candidates = [
            p for p in paras
            if (p.para_shape_id, p.style_id, p.char_shape_id_first) == chosen_triple
        ]
        rep_p = _prefer_wrapper(atom, candidates)

        # Detect body_cs (different char_shape after the first char run).
        # Used by balmun to keep "N." in number style and body in body style.
        body_cs = _detect_body_cs(rep_p, chosen_triple[2])

        # Always keep the representative paragraph — needed not just for the
        # BOX_ATOMS clone path but also for data_box (cell paragraph paraPr
        # normalization) and any future per-atom inspection.
        atom_to_slot[atom] = Slot(
            pp_id=chosen_triple[0],
            st_id=chosen_triple[1],
            cs_id=chosen_triple[2],
            template_paragraph=rep_p,
            body_cs_id=body_cs,
        )

    return SubjectCatalog(subject=subject, template=template, atom_to_slot=atom_to_slot)


# Atoms that any subject may legitimately need even if its templet doesn't
# carry an example — borrowed from `math` (the most complete catalog) when
# missing. PIC_BLOCK is the canonical case: social Q13 has a centered
# picture but social.hwpx templet doesn't include a picture-only paragraph.
_FALLBACK_ATOMS = (Atom.PIC_BLOCK, Atom.EQ_BLOCK, Atom.SEONJI_2ROW, Atom.BALMUN_CONT)


@lru_cache(maxsize=8)
def catalog_for(subject: SubjectName) -> SubjectCatalog:
    """Cached catalog. Templet rarely changes during a run.

    After building, fill in cross-subject fallback slots so atoms classified
    in src never silently drop. Subjects whose templet has no example of a
    given atom (e.g. social has no picture-only paragraph) inherit the slot
    from the math catalog, which carries every atom.
    """
    cat = _build_catalog(subject)
    if subject == "math":
        return cat
    math_cat = _build_catalog("math")
    new_slots = dict(cat.atom_to_slot)
    for atom in _FALLBACK_ATOMS:
        if atom not in new_slots and atom in math_cat.atom_to_slot:
            new_slots[atom] = math_cat.atom_to_slot[atom]
    if len(new_slots) == len(cat.atom_to_slot):
        return cat
    from dataclasses import replace as _replace
    return _replace(cat, atom_to_slot=new_slots)


__all__ = ["Slot", "SubjectCatalog", "catalog_for"]

File: kdsnr-hwp-engine/parser/src/split/compose/slot_catalog.rs
//! Derive the atom→slot table from a subject template.
//!
//! Each subject template carries one example paragraph per "slot" (balmun,
//! seonji, data box, bogi box, ...). We classify every top-level template
//! paragraph with the shared `classify_atom` and record, per atom, the
//! canonical (para_shape, style, char_shape) triple plus a representative
//! paragraph (needed to clone box shells). `apply_atom` then retargets each
//! source paragraph to its atom's slot.

use std::collections::HashMap;

use crate::model::document::Document;
use crate::model::paragraph::Paragraph;
use crate::model::style::HeadType;
use crate::split::{classify_atom, first_table, is_bogi_table, AtomKind, Subject};

/// A template slot: the paragraph-level styling an atom adopts in the output.
#[derive(Debug, Clone)]
pub struct Slot {
    pub para_shape_id: u16,
    pub style_id: u8,
    /// First char_shape of the slot (the unified body font for the atom).
    pub char_shape_id: u32,
    /// For balmun: a distinct char_shape used after the leading "N." run.
    pub body_char_shape_id: Option<u32>,
    /// Representative template paragraph (box shell clone source).
    pub template_paragraph: Option<Paragraph>,
    /// Science/social bogi: keep the source table instead of cloning the shell.
    pub preserve_source_table: bool,
    /// Science/social: flatten tiny 1x1 marker tables inside cells back to text.
    pub flatten_cell_marker_tables: bool,
}

/// The atom→slot table for one subject.
#[derive(Debug, Clone, Default)]
pub struct SubjectCatalog {
    pub atom_to_slot: HashMap<AtomKind, Slot>,
}

fn char_shape_first(p: &Paragraph) -> u32 {
    p.char_shapes.first().map(|c| c.char_shape_id).unwrap_or(0)
}

/// First run whose char_shape differs from the paragraph's first run — the
/// balmun template's body style (the leading "N." run uses the first style).
fn body_char_shape(p: &Paragraph, first: u32) -> Option<u32> {
    p.char_shapes
        .iter()
        .map(|c| c.char_shape_id)
        .find(|&id| id != first)
}

/// Build the atom→slot table from a subject template. `preserve_source_bogi`
/// keeps the source bogi table verbatim (and flattens cell marker tables)
/// instead of cloning the template's box shell.
pub fn build_catalog(
    template: &Document,
    subject: Subject,
    preserve_source_bogi: bool,
) -> SubjectCatalog {
    let empty: Vec<Paragraph> = Vec::new();
    let body = template
        .sections
        .first()
        .map(|s| &s.paragraphs)
        .unwrap_or(&empty);

    let mut by_atom: HashMap<AtomKind, Vec<&Paragraph>> = HashMap::new();
    let mut prev = None;
    for p in body {
        let atom = if subject == Subject::Korean && is_numbered_paragraph(template, p) {
            AtomKind::Balmun
        } else {
            classify_atom(p, subject, prev)
        };
        if matches!(atom, AtomKind::Empty | AtomKind::Unknown) {
            continue;
        }
        prev = Some(atom);
        by_atom.entry(atom).or_default().push(p);
    }

    let mut atom_to_slot = HashMap::new();
    for (atom, paras) in by_atom {
        // Most-common (para_shape, style, char) triple for this atom.
        let mut counts: HashMap<(u16, u8, u32), usize> = HashMap::new();
        for p in &paras {
            *counts
                .entry((p.para_shape_id, p.style_id, char_shape_first(p)))
                .or_default() += 1;
        }
        let chosen = counts
            .iter()
            .max_by_key(|(_, c)| **c)
            .map(|(k, _)| *k)
            .unwrap_or((0, 0, 0));

        // Representative paragraph for cloning: for bogi boxes prefer a clean
        // bogi-marked wrapper (table only, no fused text); otherwise the first
        // paragraph matching the chosen triple.
        let rep = if atom == AtomKind::BogiBox {
            paras
                .iter()
                .copied()
                .find(|p| first_table(p).is_some_and(is_bogi_table) && p.text.trim().is_empty())
                .or_else(|| {
                    paras
                        .iter()
                        .copied()
                        .find(|p| first_table(p).is_some_and(is_bogi_table))
                })
                .unwrap_or(paras[0])
        } else {
            paras
                .iter()
                .copied()
                .find(|p| (p.para_shape_id, p.style_id, char_shape_first(p)) == chosen)
                .unwrap_or(paras[0])
        };
        let mut slot_char_shape = chosen.2;
        let mut balmun_body_char_shape = body_char_shape(rep, chosen.2);
        if atom == AtomKind::Balmun && balmun_body_char_shape.is_none() {
            if let Some(number_cs) = numbering_head_char_shape(template, rep) {
                slot_char_shape = number_cs;
                balmun_body_char_shape = Some(chosen.2);
            }
        }

        atom_to_slot.insert(
            atom,
            Slot {
                para_shape_id: chosen.0,
                style_id: chosen.1,
                char_shape_id: slot_char_shape,
                body_char_shape_id: balmun_body_char_shape,
                template_paragraph: Some(rep.clone()),
                preserve_source_table: preserve_source_bogi && atom == AtomKind::BogiBox,
                flatten_cell_marker_tables: preserve_source_bogi,
            },
        );
    }

    SubjectCatalog { atom_to_slot }
}

fn numbering_head_char_shape(template: &Document, p: &Paragraph) -> Option<u32> {
    let para_shape = template
        .doc_info
        .para_shapes
        .get(p.para_shape_id as usize)?;
    if para_shape.head_type != HeadType::Number || para_shape.numbering_id == 0 {
        return None;
    }
    let numbering = template
        .doc_info
        .numberings
        .get(para_shape.numbering_id.saturating_sub(1) as usize)?;
    let char_shape_id = numbering
        .heads
        .get(para_shape.para_level as usize)
        .map(|head| head.char_shape_id)?;
    (char_shape_id != u32::MAX).then_some(char_shape_id)
}

fn is_numbered_paragraph(template: &Document, p: &Paragraph) -> bool {
    template
        .doc_info
        .para_shapes
        .get(p.para_shape_id as usize)
        .is_some_and(|para_shape| para_shape.head_type == HeadType::Number)
}
